event-publisher-spi/src/test/scala/uk.gov.nationalarchives.eventpublisherspi/UserMonitoringTaskSpec.scala

```diff
@@ -0,0 +1,141 @@
+    when(mockRealmProvider.getRealmsStream).thenReturn(java.util.stream.Stream.of(mockRealmModel))
+    when(mockSession.users()).thenReturn(mockUserProvider)
+    when(mockSession.realms()).thenReturn(mockRealmProvider)
+    when(mockSession.userCredentialManager()).thenReturn(mockUserCredentialManager)
+    new UserMonitoringTask(mockSnsUtils, EventPublisherConfig("https://example.com", "", "test"), credentialType).run(mockSession)
+
+    verifyZeroInteractions(mockSnsUtils)
+  }
+
+  "The run method" should "send a message if there is one user with MFA and one without" in {
+    val mockSession = mock[KeycloakSession]
+    val mockRealmProvider = mock[RealmProvider]
+    val mockRealmModel = mock[RealmModel]
+    val mockUserCredentialManager = mock[UserCredentialManager]
+    val mockUserModelWithMFA = mock[UserModel]
+    val mockUserModelWithoutMFA = mock[UserModel]
+    val mockUserProvider = mock[UserProvider]
+    val mockSnsUtils = mock[SNSUtils]
+    val userId = "dfd7356c-e6e0-40dd-affd-de04e29ad359"
+    val topicArn = "snsTopicArn"
+    val credentialType = "otp"
+
+    when(mockRealmModel.getName).thenReturn("testRealm")
+    when(mockUserModelWithoutMFA.getId).thenReturn(userId)
+    when(mockUserProvider.getUsersStream(mockRealmModel)).thenReturn(java.util.stream.Stream.of(mockUserModelWithMFA, mockUserModelWithoutMFA))
+    when(mockUserCredentialManager.isConfiguredFor(mockRealmModel, mockUserModelWithMFA, credentialType)).thenReturn(true)
+    when(mockUserCredentialManager.isConfiguredFor(mockRealmModel, mockUserModelWithoutMFA, credentialType)).thenReturn(false)
+    when(mockRealmProvider.getRealmsStream).thenReturn(java.util.stream.Stream.of(mockRealmModel))
+    when(mockSession.users()).thenReturn(mockUserProvider)
+    when(mockSession.realms()).thenReturn(mockRealmProvider)
+    when(mockSession.userCredentialManager()).thenReturn(mockUserCredentialManager)
+    new UserMonitoringTask(mockSnsUtils, EventPublisherConfig("https://example.com", topicArn, "test"), credentialType).run(mockSession)
+
+    val expectedMessage = """{
+                             |  "tdrEnv" : "test",
+                             |  "message" : "\ntestRealm realm has 1 user without MFA\ndfd7356c-e6e0-40dd-affd-de04e29ad359\n"
+                             |}""".stripMargin
+
+    verify(mockSnsUtils, times(1)).publish(expectedMessage, topicArn)
+  }
+
+  "The run method" should "send a message if there are two users with MFA missing" in {
+    val mockSession = mock[KeycloakSession]
+    val mockRealmProvider = mock[RealmProvider]
+    val mockRealmModel = mock[RealmModel]
+    val mockUserCredentialManager = mock[UserCredentialManager]
+    val mockUserModelWithoutMFA1 = mock[UserModel]
+    val mockUserModelWithoutMFA2 = mock[UserModel]
+    val mockUserProvider = mock[UserProvider]
+    val mockSnsUtils = mock[SNSUtils]
+    val userId1 = "dfd7356c-e6e0-40dd-affd-de04e29ad359"
```

Comment:
User ids used in multiple tests. Are they worth abstracting out into class vals?

Reply:
I've moved all the user IDs as well as the topic arn and the credential type.